### ISSUE TITLE
Training QuadX-Pole-Balance-v4 leads to graceful landings of drone

### ISSUE THREAD

**Issue by earlzero** [jjshoots/PyFlyt#94]
Hi

Thank you for the library, it is very easy to start building stuff with it. I have a question about environment "PyFlyt/QuadX-Pole-Balance-v4". I've tried to use PPO from stable_baselines3 for this environment for up to 4000000 steps.

train_loss falls to 0.0055, value_loss to 4.19e-6 so it seems that training succeeded. The behavior of drone I'm seein is very graceful landing. I'd expect drone to remaing flying for quite a while, instead it always land.

More training I go, more gracefully it lands :).

I initially assumed that I might undetrained the model, but adding 2000000 (to initial 2000000 steps) didn't help. I'm curious which algorithm was used to produce [demo](https://taijunjet.com/PyFlyt/documentation/gym_envs/quadx_pole_waypoints_env.html).

I had a theory that reward function doesn't penalize this behavior, but not sure about it.

Thanks!

```
import sys
from collections import defaultdict
import gymnasium as gym
import numpy as np
from tqdm import tqdm  # Progress bar
from stable_baselines3 import A2C
from stable_baselines3.common.env_util import make_vec_env
from gymnasium.wrappers import FlattenObservation
from stable_baselines3.common.vec_env import DummyVecEnv
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import SubprocVecEnv
from typing import Callable
import PyFlyt.gym_envs


def learning_schedule(initial_value: float) -> Callable[[float], float]:
    def func(progress_remaining: float) -> float:
        if progress_remaining < 0.25:
            return initial_value / 8
        elif progress_remaining < 0.5:
            return initial_value / 4
        elif progress_remaining < 0.75:
            return initial_value / 2
        else:
            return initial_value
    return func

if __name__ == '__main__':
    def prepare_hover_env(render_mode):
        def make_hover_env():
            env = gym.make("PyFlyt/QuadX-Pole-Balance-v4", render_mode=render_mode) 
            env = FlattenObservation(env)
            return env
        return make_hover_env

    vec_env = SubprocVecEnv([prepare_hover_env(None) for i in range(4)]) 

    model = PPO("MlpPolicy", vec_env, verbose=1, tensorboard_log="./a2c_cartpole_tensorboard/", learning_rate=learning_schedule(0.0012),device="cpu")
    model.learn(total_timesteps=4000000)
    model.save("ppo_pole")
    del model

    vec_env = SubprocVecEnv([prepare_hover_env("human") for i in range(1)]) 
    model = A2C.load("ppo_pole")

    obs = vec_env.reset()
    total_rewards = 0
    count_rewards = 0
    while True:
        action, _states = model.predict(obs)
        obs, rewards, dones, info = vec_env.step(action)
        vec_env.render("human")
        total_rewards+=rewards
        count_rewards+=1
```

**Comment by jjshoots** [comment]
Thanks for the issue! That's a very interesting learned behavior. What's likely happening is the agent is prioritizing future rewards more and afraid of dropping the pole -- but this is just speculation. Have you tried reducing the discount factor?

The algorithm used to train the demo was SAC, using [these](https://github.com/jjshoots/RLTeamAerialCombat/blob/main/src/configs/quadx_pole_balance_config.yaml) hyperparameters.

**Comment by earlzero** [comment]
Thank you, this was helpful!

First of all I was saving PPO policy, but loading A2C policy (yet it even produced reasonable results - drone landing).

When I fixed this bug I still had the same behavior independently of level of discount factor (I tried 0.9-0.99), but when I switched to SAC, it worked even without discount factor.

I will try to make PPO work anyway and leave a comment here later with the results (might be helpful to someone else trying PyFlyt).

**Comment by earlzero** [comment]
It took me way more time than I expected to figure it out. Default hyperparameters were they off for PPO and important ones ended up being:
* `ent_coef` - I used 0.01 (default 0)
* `gamma`. I used 0.95 (default 0.99)
* `learning_rate`. I run multiple experiments with decaying learning rate, but in the end 5e-4 was sufficient.

It needed 14 millions of steps to get decent results (drone hovering for over a minute).

I think my parameters are still not great, so with more tuning it should be possible to learn faster than 14 millions steps.

Here is example code I created
```
import gymnasium as gym
import numpy as np
from tqdm import tqdm  # Progress bar
from stable_baselines3 import A2C
from stable_baselines3.common.env_util import make_vec_env
from gymnasium.wrappers import FlattenObservation
from stable_baselines3.common.vec_env import DummyVecEnv, VecNormalize
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import SubprocVecEnv
from typing import Callable
import PyFlyt.gym_envs
from stable_baselines3.common.callbacks import (
    CheckpointCallback,
    EvalCallback,
    BaseCallback,
)

def learning_schedule(initial_value: float) -> Callable[[float], float]:
    def func(progress_remaining: float) -> float:
        return max(progress_remaining * initial_value, 1e-6)

    return func


def learning_schedule_exp(initial_value: float) -> Callable[[float], float]:
    def func(progress_remaining: float) -> float:
        return (1 - 0.9**progress_remaining) * 10 * initial_value

    return func


if __name__ == "__main__":

    def prepare_hover_env(render_mode):
        def make_hover_env():
            env = gym.make("PyFlyt/QuadX-Pole-Balance-v4", render_mode=render_mode)
            env = FlattenObservation(env)
            return env

        return make_hover_env

    vec_env = SubprocVecEnv([prepare_hover_env(None) for i in range(8)])

    eval_callback = EvalCallback(
        vec_env,
        callback_on_new_best=None,
        n_eval_episodes=5,
        best_model_save_path=".",
        log_path="logs",
        eval_freq=10000,
    )

    checkpoint_callback = CheckpointCallback(
        save_freq=100000,
        save_path="./logs/",
        name_prefix="rl_model",
        save_replay_buffer=True,
        save_vecnormalize=True,
    )

    model = PPO(~
        "MlpPolicy",
        vec_env,
        verbose=0,  
        tensorboard_log="./pole_balance/",
        device="cpu",
        ent_coef=0.01,
        gamma=0.95,
        learning_rate=5e-4,
    )

    model.learn(total_timesteps=30000000, callback=[checkpoint_callback, eval_callback])
    model.save("ppo_pole")
    del model

    vec_env = SubprocVecEnv([prepare_hover_env("human") for i in range(1)])
    model = PPO.load("ppo_pole")

    obs = vec_env.reset()
    total_rewards = 0
    count_rewards = 0
    while True:
        action, _states = model.predict(obs)
        obs, rewards, dones, info = vec_env.step(action)
        vec_env.render("human")
        total_rewards += rewards
        count_rewards += 1
```
